### PR TITLE
[DSS-153] Button Web Component

### PR DESCRIPTION
**Pull request by goodwinchris**
# Description

Create MVP of Sage-Button with related stories, documentation, and tests

Fixes [DSS-153 - Develop "Button" Web Component](https://kajabi.atlassian.net/browse/DSS-153)

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

- [x] unit tests
- [x] e2e tests
- [x] accessibility tests
- [x] tested manually

**Test Configuration**:

- sage versions:
- os:
- browsers:
- screen readers:
- misc:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes

### CODE REVIEW

File: libs/core/src/components/my-component/stories/my-component.stories.mdx [hunk 1]
```diff
@@ -16,7 +16,7 @@ This is a sample of using MDX with Web Components.
 <ArgsTable of="my-component" />
 
 <Canvas>
-  <Story name="Primary" args={{ first: 'Aldo', middle: 'Charlie', last: 'Yates' }}>
+  <Story name="Primary" args={{ first: 'Aldo', middle: '"Sturdy Dollop"', last: 'Yates' }}>
```
Comment (ju-Skinner) [inline libs/core/src/components/my-component/stories/my-component.stories.mdx]: That middle name is 🔥 😂

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+   * When set to true, icon is added to button
+   * @defaultValue false
+   */
+  @Prop() icon: string;
+  
+  /**
+   * When set to true, buttons are disabled
+   * @defaultValue false
+   */
+  @Prop({ reflect: true }) disabled: true;
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/sage-button.tsx]: We had a convo about `reflect` in Slack. I'll send you the thread.

File: libs/core/src/components/sage-button/test/sage-button.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    const page = await newSpecPage({
+      components: [SageButton],
+      html: `<sage-button></sage-button>`,
+    });
+    expect(page.root).toEqualHtml(`
+      <sage-button>
+        <mock:shadow-root>
+          <slot></slot>
+        </mock:shadow-root>
+      </sage-button>
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/test/sage-button.spec.tsx]: Don't be afraid to fix the failing test. It's fairly simple. Think about the _**default**_ output based on what is in the `html: ` attribute.


The answer is in the output 😉

**Comment by ju-Skinner** [comment]
Great start 💯

File: libs/core/src/components/sage-button/sage-button.tsx (hunk 2 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-button/sage-button.tsx]: Sounds like we likely don't need this right now.

File: libs/core/src/components/sage-button/test/sage-button.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    expect(root).toEqualHtml(`
+      <sage-button variant="primary">
+        <mock:shadow-root>
+          <button class="sage-button sage-button--primary">
+            <slot></slot>
+          </button>
+        </mock:shadow-root>
+      </sage-button>
+    `);
+  });
```
Comment (goodwinchris) [inline libs/core/src/components/sage-button/test/sage-button.spec.tsx]: Un-needed

File: libs/core/src/components/sage-button/test/sage-button.e2e.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+  it('renders when slot is used', async () => {
+    const page = await newE2EPage();
+    await page.setContent('<sage-button>Test Content</sage-button>');
+    const component = await page.find('sage-button');
+    expect(component).toHaveClass('hydrated');
+
+    const element = await page.find('sage-button');
+    // Question: When should I use `toBe` and when should I use `toEqual`?
+    expect(element.textContent).toBe('Test Content');
```
Comment (goodwinchris) [inline libs/core/src/components/sage-button/test/sage-button.e2e.ts]: Check for text version

File: libs/core/src/components/sage-button/test/sage-button.e2e.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    const component = await page.find('sage-button');
+    expect(component).toHaveClass('hydrated');
+
+    const element = await page.find('sage-button >>> button');
+    let icon = await element.find('svg');
+    expect(icon).toBeNull();
+
+    component.setProperty('icon', 'trashIcon');
+    await page.waitForChanges();
+    icon = await element.find('svg');
```
Comment (goodwinchris) [inline libs/core/src/components/sage-button/test/sage-button.e2e.ts:92]: Make sure we check for SVG path

File: libs/core/src/components/sage-button/test/sage-button.e2e.ts (hunk 5 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-button/test/sage-button.e2e.ts]: Follow up: using `toMatch()` https://jestjs.io/docs/using-matchers#strings

**Comment by goodwinchris** [comment]
Look into onclick addition to button

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+          fakeButton.type = this.type
+          fakeButton.style.display = 'none'
+          form.appendChild(fakeButton)
+          fakeButton.click()
+          fakeButton.remove()
+        }
+      }
+    } else {
+      // If button clicked IS associated with a form
+    }
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/sage-button.tsx]: Since we aren't doing anything inside of the else, remove it for now.

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  
+  private buttonClassNames = () => {
+    let className = `sage-button`;
+    // console.log(className);
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/sage-button.tsx]: Please remove commented code

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    if (this.variant && this.variant != 'primary') {
+      const variantClassName = `sage-button--${this.variant}`;
+      className += ' ' + variantClassName;
+    }
+    // return className;
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/sage-button.tsx]: same as above

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    return className;
+    // console.log(className);
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/sage-button.tsx]: same as above

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+      <h3>When to use a Link</h3>
+    </td>
+  </tr>
+  <tr>
+    <td>
+      To trigger on-page interactivity <br /> To submit a form
+    </td>
+    <td>To send a user to another URL or anchor</td>
+  </tr>
+</table>
```
Comment (ju-Skinner) [inline libs/core/src/components/sage-button/stories/sage-button.stories.mdx]: I would like to see more detail here. Remember we discussed err'ing more on the verbose side.

File: libs/core/src/global/styles/_variables.scss [hunk 12]
```diff
@@ -0,0 +1,5 @@
+////
+/// Global variables
+///
+/// @group sage
+////
```
Comment (ju-Skinner) [inline libs/core/src/global/styles/_variables.scss]: Is this file still needed?

File: libs/core/src/components/sage-button/sage-button.tsx [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+    let className = `sage-button`;
+    if (this.variant && this.variant != 'primary') {
+      const variantClassName = `sage-button--${this.variant}`;
+      className += ' ' + variantClassName;
+    }
+    return className;
+  };
+
+  render() {
+    const trashIcon = (
```
Comment (anechol) [inline libs/core/src/components/sage-button/sage-button.tsx:73]: Something about this icon is throwing off the button alignment. It's extremely subtle, but if you squint you notice it in the stories.

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+- Directing users to a new page or different part within the same page. Instead, use [sage-link](/docs/webcomponents-sage-link--external).
+- Limited space available. Instead consider an IconButton.
+
+## Properties
+
+<ArgsTable of="sage-button" />
+
+## Variant
+
+### Default (Primary)
```
Comment (anechol) [inline libs/core/src/components/sage-button/stories/sage-button.stories.mdx:42]: Cool! I like seeing the different button variations next to each other.

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+<Meta title="webcomponents/sage-button" component="sage-button" argTypes={extractArgTypes('sage-button')} />
+
+export const Template = (args) =>
+  html` <sage-button value=${args.value} variant=${args.variant} name=${args.name} icon=${args.icon} disabled=${args.disabled} type=${args.type}> ${args.slot} </sage-button> `;
+
+# sage-button
+
+A button component to generate all versions of `sage-button` as outlined in [Figma](https://www.figma.com/file/sMbtLUHSt2vfKgKjyQ3052/%F0%9F%A7%A9-Sage-components?node-id=3%3A15&t=dIcvX1VtOu0FL0VE-0).
+
+## Usage Guidelines
```
Comment (anechol) [inline libs/core/src/components/sage-button/stories/sage-button.stories.mdx:15]: The only thing I can think of to add here could be some guidelines and best practices around accessibility, but again, we can always revisit docs when we ave our retro.

**Review by anechol — COMMENTED** [review]
Nice job! Just a few comments.

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+This is our default button type.
+
+### Submit
+
+The button submits the form data to the server.
+
+### Reset
+
+The button resets all the controls to their initial values. (This behavior tends to annoy users.)
```
Comment (QuintonJason) [inline libs/core/src/components/sage-button/stories/sage-button.stories.mdx]: `(This behavior tends to annoy users.)`
Do you have an accessibility article or ux documentation to support this claim? I don't think this portion of documentation should be opinionated.

File: package.json [hunk 17]
```diff
@@ -32,5 +32,6 @@
     "commitizen": {
       "path": "./node_modules/cz-conventional-changelog"
     }
-  }
+  },
+  "dependencies": {}
```
Comment (QuintonJason) [inline package.json]: This will probably disappear after a rebase.

File: libs/core/src/components/sage-link/readme.md [hunk 18]
```diff
@@ -4,6 +4,7 @@ Link is mainly used as navigational element and usually appear within or directl
```
Comment (QuintonJason) [inline libs/core/src/components/sage-link/readme.md]: These file updates can be removed

File: libs/core/src/components/sage-button/test/sage-button.e2e.ts [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+  });
+
+  it('renders when slot is used', async () => {
+    const page = await newE2EPage();
+    await page.setContent('<sage-button>Test Content</sage-button>');
+    const component = await page.find('sage-button');
+    expect(component).toHaveClass('hydrated');
+
+    const element = await page.find('sage-button');
+    // Question: When should I use `toBe` and when should I use `toEqual`?
```
Comment (QuintonJason) [inline libs/core/src/components/sage-button/test/sage-button.e2e.ts]: Is this comment still needed?

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx (hunk 16 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-button/stories/sage-button.stories.mdx]: I'll push a commit with a link to MDN but since MDN calls it out, I figured its valuable for us to note it in our docs as well. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#attr-type

File: libs/core/src/components/sage-button/test/sage-button.e2e.ts (hunk 19 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-button/test/sage-button.e2e.ts]: Nope, good catch!

File: package.json (hunk 17 above)
Comment (goodwinchris) [reply package.json]: Had to manually remove for some reason but I'll push that up in a new commit.

File: libs/core/src/components/sage-link/readme.md (hunk 18 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-link/readme.md]: I'll revert this back, not sure why it changed.

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx (hunk 15 above)
Comment (goodwinchris) [reply libs/core/src/components/sage-button/stories/sage-button.stories.mdx:15]: Added a section, feel free to note anything you think could be included other than what I've added!

**Review by ju-Skinner — APPROVED** [review]
👏🏾 Good job

File: libs/core/src/components/sage-button/stories/sage-button.stories.mdx (hunk 16 above)
Comment (QuintonJason) [reply libs/core/src/components/sage-button/stories/sage-button.stories.mdx]: If they call it out, I suppose it's fine. I'm confused by that wording from them because the intent of the `reset` property is to reset all the form fields. Not sure how that would annoy users unless it's not being used correctly. But alas,

**Review by QuintonJason — APPROVED** [review]
Good to go

**Review by anechol — APPROVED** [review]
Awesome work on this! Just a minor nitpick/question:

This default button looks disabled. Should we be passing `false` to the disabled prop here? And a further question is do we need to always provide `false` for this to render correctly or is this just a Storybook issue?
Screen Shot 2023-01-17 at 11 06 01 AM

**Review by anechol — APPROVED** [review]
